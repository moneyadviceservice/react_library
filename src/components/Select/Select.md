The Select input allows a user to select a value from a series of options.

```jsx
import React, { useState } from 'react';
import { Form, Formfield, Select, Paragraph as P } from '@moneypensionservice/directories';

const [animal, changeAnimal] = useState('🐕 Dog');

<>
<Form>
  <Formfield>
    <Select
      disabled
      id="disabled"
      label="I am disabled!"
      options={['Try to select me 🙈']}
    />
  </Formfield>
  <Formfield>
    <Select
      id="animals"
      label="Select your favourite animal:"
      onChange={e => changeAnimal(e.target.value)}
      options={[
        '🐕 Dog',
        '🦊 Fox',
        '🐈 Cat',
        '🐎 Horse',
      ]}
    />
  </Formfield>
</Form>
<P margin={{top: '10px'}}>{`Currently selected: ${animal}`}</P>
</>

```
