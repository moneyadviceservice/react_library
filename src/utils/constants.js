export const breakpoints = {
  xs: 1,
  sm: 48, // 768px
  md: 64, // 1024px
  lg: 90, // 1440px
  xl: 120, // 1920px
}

export const dimensions = ['xs', 'sm', 'md', 'lg', 'xl']

export const mediaQuery = 'only screen'

export const ALIGN_SELF_MAP = {
  center: 'center',
  end: 'flex-end',
  start: 'flex-start',
  stretch: 'stretch',
}

export const SPACING_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
]

