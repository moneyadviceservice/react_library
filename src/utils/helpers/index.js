export * from './background'
export * from './responsive'
export * from './flexbox'
export * from './styles'
export * from './mediaQueries'
export * from './sizing'
export * from './refs'
