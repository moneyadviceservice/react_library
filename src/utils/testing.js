import { render, act } from '@testing-library/react'
import GlobalTheme from '../components/ThemeProvider'

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalTheme, ...options })

const waitRender = async (ui, options) => {
  const { container, ...rest } = customRender(ui, options)

  await act(async () => {
    await Promise.resolve(container)
  })

  return { container, ...rest }
}

export * from '@testing-library/react'

export { customRender as render, waitRender }
