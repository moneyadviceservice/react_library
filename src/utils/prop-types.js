import PropTypes from 'prop-types'
import { SPACING_SIZES } from './constants'

/** spacing prop.type */
const spacingProp = PropTypes.shape({
  bottom: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  horizontal: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  left: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  right: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  top: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  vertical: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
})

export const genericPropTypes = {
  a11yTitle: PropTypes.string,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  margin: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  padding: PropTypes.oneOfType([
    PropTypes.oneOf([...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  border: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
}

export function genericPropsDefaults({
  a11yTitle = null,
  alignSelf = null,
  margin = null,
  padding = null,
  border = null,
} = {}) {
  return {
    a11yTitle: a11yTitle,
    alignSelf: alignSelf,
    margin: margin,
    padding: padding,
    border: border,
  }
}
