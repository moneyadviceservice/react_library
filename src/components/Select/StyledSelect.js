import styled, { css } from 'styled-components'
import { genericStyles } from '../../utils/helpers'

const Label = styled.label`
  ${({ disabled, theme }) =>
    disabled &&
    css`
      color: ${theme.colors.disabledText};
    `}
  ${({ hide }) =>
    hide
      ? css`
          display: none;
        `
      : css`
          display: block;
        `}

  padding-bottom: 0.5rem;
`

const Field = styled.select`
  ${genericStyles}

  ${({ margin }) =>
    !margin &&
    css`
      margin-bottom: 12px;
    `}
  ${({ padding }) =>
    !padding &&
    css`
      padding: 3px;
    `}
  ${({ width }) =>
    css`
      width: ${width};
    `}
  
  height: 30px;
  text-transform: none;
`

export { Label, Field }
