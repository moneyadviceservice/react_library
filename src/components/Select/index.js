import React from 'react'
import PropTypes from 'prop-types'
import { genericPropTypes, genericPropsDefaults } from '../../utils/prop-types'
import { Label, Field } from './StyledSelect'

const Select = ({
  a11yTitle,
  children,
  disabled,
  hideLabel,
  id,
  label,
  width,
  onChange,
  options,
  ...rest
}) => (
  <>
    {label && (
      <Label disabled={disabled} hide={hideLabel} htmlFor={id}>
        {label}
      </Label>
    )}
    <Field
      width={width}
      aria-label={a11yTitle}
      disabled={disabled}
      id={id}
      onChange={onChange}
      {...rest}
    >
      {options.length &&
        options.map((option, index) => (
          <option key={index} value={option}>
            {option}
          </option>
        ))}
      {children}
    </Field>
  </>
)

// Documentation
Select.propTypes = {
  /** Add options to the select element without using the provided prop. */
  children: PropTypes.any,
  /** Whether the select element is disabled. */
  disabled: PropTypes.bool,
  /** Hides the select input label. */
  hideLabel: PropTypes.bool,
  /** Id of the select input. Also sets the "for" attribute on the label. Required. */
  id: PropTypes.string.isRequired,
  /** Caption for the select element. */
  label: PropTypes.string,
  /** Sets the width of the select element. */
  width: PropTypes.string,
  /** Options for the select element. */
  options: PropTypes.array.isRequired,
  /** On change trigger function event. */
  onChange: PropTypes.func,
  ...genericPropTypes,
}

Select.defaultProps = {
  disabled: false,
  hideLabel: false,
  ...genericPropsDefaults(),
}

/** @component */
export { Select }
