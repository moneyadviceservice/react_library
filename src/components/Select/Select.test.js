import React from 'react'
import { render, fireEvent } from '../../utils/testing'
import { Select } from '../Select'

describe('<Select />', () => {
  const options = ['option1', 'option2', 'option3']

  describe('Snapshots', () => {
    it('should match snapshot with options, id and label', () => {
      const { container } = render(
        <Select id="test" label="test label" options={['option 1']} />
      )
      expect(container).toMatchSnapshot()
    })
    it('should match snapshot disabled', () => {
      const { container } = render(
        <Select
          disabled
          id="disabled"
          label="test disabled"
          options={['option 1']}
        />
      )
      expect(container).toMatchSnapshot()
    })
    it('should match snapshot with multiple options', () => {
      const { container } = render(
        <Select id="options" label="test options" options={options} />
      )
      expect(container).toMatchSnapshot()
    })
  })

  describe('options prop', () => {
    it('should display correct number of options', () => {
      const { container } = render(<Select id="options" options={options} />)
      const dropDownOptions = container.querySelectorAll('option')

      expect(dropDownOptions.length).toBe(3)
    })

    it('should display correct option text', () => {
      const { container } = render(<Select id="options" options={options} />)
      const dropDownOptions = container.querySelectorAll('option')
      const [opt1, opt2, opt3] = Array.from(dropDownOptions)

      expect(opt1.textContent).toBe(options[0])
      expect(opt2.textContent).toBe(options[1])
      expect(opt3.textContent).toBe(options[2])
    })
  })

  describe('onChange prop', () => {
    it('should call onChange function when option is changed', () => {
      const onChangeMock = jest.fn()
      const { getByTestId } = render(
        <Select
          data-testid="select"
          id="change"
          onChange={onChangeMock}
          options={options}
        />
      )

      fireEvent.change(getByTestId('select'), {
        target: { value: 'option2' },
      })

      expect(onChangeMock).toHaveBeenCalledTimes(1)
    })
  })
})
