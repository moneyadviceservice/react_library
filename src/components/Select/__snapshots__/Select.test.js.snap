// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> Snapshots should match snapshot disabled 1`] = `
<div>
  <label
    class="StyledSelect__Label-sc-17qnzyp-0 kOMahN"
    disabled=""
    for="disabled"
  >
    test disabled
  </label>
  <select
    class="StyledSelect__Field-sc-17qnzyp-1 hGHanr"
    disabled=""
    id="disabled"
  >
    <option
      value="option 1"
    >
      option 1
    </option>
  </select>
</div>
`;

exports[`<Select /> Snapshots should match snapshot with multiple options 1`] = `
<div>
  <label
    class="StyledSelect__Label-sc-17qnzyp-0 feVND"
    for="options"
  >
    test options
  </label>
  <select
    class="StyledSelect__Field-sc-17qnzyp-1 hGHanr"
    id="options"
  >
    <option
      value="option1"
    >
      option1
    </option>
    <option
      value="option2"
    >
      option2
    </option>
    <option
      value="option3"
    >
      option3
    </option>
  </select>
</div>
`;

exports[`<Select /> Snapshots should match snapshot with options, id and label 1`] = `
<div>
  <label
    class="StyledSelect__Label-sc-17qnzyp-0 feVND"
    for="test"
  >
    test label
  </label>
  <select
    class="StyledSelect__Field-sc-17qnzyp-1 hGHanr"
    id="test"
  >
    <option
      value="option 1"
    >
      option 1
    </option>
  </select>
</div>
`;
